### PR TITLE
Modularize logging components

### PR DESCRIPTION
**Pull request by felipecrescencio-cit**
Fix issue 674.
PR changes:

- New inline module that centralizes logging configuration based on SDW module and adapted to support projects, folders, and organizations
- It supports all logging destinations: bigquery, storage, pubsub, and logbucket
- Support to logbucket destination was added on May 31 on `log-export` module and now it is supported by the example foundation
- Changed logging in `1-org` to use the new module

### CODE REVIEW

File: 1-org/envs/shared/outputs.tf [hunk 1]
```diff
@@ -93,3 +93,8 @@ output "logs_export_storage_bucket_name" {
   value       = module.storage_destination.resource_name
   description = "The storage bucket for destination of log exports"
 }
+
+output "logs_export_logbucket_name" {
+  value       = module.logbucket_destination.resource_name
+  description = "The log bucket for destination of log exports"
```
Comment (daniel-cit) [inline 1-org/envs/shared/outputs.tf]:
```suggestion
  description = "The log bucket for destination of log exports. See https://cloud.google.com/logging/docs/routing/overview#buckets"
```

File: 1-org/modules/centralized-logging/README.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
+# Centralized Logging Module
+
+This module handles logging configuration enabling one or more resources such as organization, folders, or projects to send logs to a destination: GCS bucket, Big Query, Pub/Sub, or Log Bucket.
```
Comment (daniel-cit) [inline 1-org/modules/centralized-logging/README.md]:
```suggestion
This module handles logging configuration enabling one or more resources such as organization, folders, or projects to send logs to a destination: [GCS bucket](https://cloud.google.com/logging/docs/export/using_exported_logs#gcs-overview), [Big Query](https://cloud.google.com/logging/docs/export/bigquery), [Pub/Sub](https://cloud.google.com/logging/docs/export/using_exported_logs#pubsub-overview), or [Log Buckets](https://cloud.google.com/logging/docs/routing/overview#buckets).
```

File: 1-org/modules/centralized-logging/README.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  include_children               = true
+  logging_target_type            = "storage"
+  logging_destination_project_id = "<log_destination_project_id>"
+  logging_target_name            = "bkt-audit-logs"
+  uniform_bucket_level_access    = true
+  logging_location               = "US"
+}
+```
+
+Heads up when the destination is a Log Bucket and the logging destination project is also a resource. If it is the case, do not forget to set `logging_project_key` variable with the logging destination project key from map resources. Get more details at [Configure and manage sinks](https://cloud.google.com/logging/docs/export/configure_export_v2#dest-auth:~:text=If%20you%27re%20using%20a%20sink%20to%20route%20logs%20between%20Logging%20buckets%20in%20the%20same%20Cloud%20project%2C%20no%20new%20service%20account%20is%20created%3B%20the%20sink%20works%20without%20the%20unique%20writer%20identity.).
```
Comment (daniel-cit) [inline 1-org/modules/centralized-logging/README.md]:
```suggestion
**Note:** when the destination is a Log Bucket and a sink is been created in the same project, set variable 
 `logging_project_key` with the **key** used to map the Log Bucket project in the `resources` map.
Get more details at [Configure and manage sinks](https://cloud.google.com/logging/docs/export/configure_export_v2#dest-auth:~:text=If%20you%27re%20using%20a%20sink%20to%20route%20logs%20between%20Logging%20buckets%20in%20the%20same%20Cloud%20project%2C%20no%20new%20service%20account%20is%20created%3B%20the%20sink%20works%20without%20the%20unique%20writer%20identity.).
```

File: 1-org/modules/centralized-logging/main.tf [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  logging_sink_name          = length(var.logging_sink_name) > 0 ? var.logging_sink_name : "sk-to-${local.logging_target_name_prefix}-${var.logging_destination_project_id}"
+  logging_target_name_prefix = var.logging_target_type == "bigquery" ? "ds" : var.logging_target_type == "pubsub" ? "topic" : var.logging_target_type == "storage" ? "bkt" : "logbkt"
+  logging_target_name        = length(var.logging_target_name) > 0 ? var.logging_target_name : "${local.logging_target_name_prefix}-${random_string.suffix.result}"
+  log_exports                = setunion(local.log_exports_others)
+  parent_resource_ids        = [for parent_resource_id in local.log_exports[*].parent_resource_id : parent_resource_id]
+
+  # Bigquery sink options
+  bigquery_options = var.logging_target_type == "bigquery" && var.bigquery_options != null ? var.bigquery_options : null
+
+  log_exports_others = toset([
```
Comment (daniel-cit) [inline 1-org/modules/centralized-logging/main.tf]: adding a comment explaining why we need this could be useful

File: 1-org/modules/centralized-logging/main.tf [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    for value in module.log_export : value
+  ])
+}
+
+resource "random_string" "suffix" {
```
Comment (daniel-cit) [inline 1-org/modules/centralized-logging/main.tf:54]: why are you adding another random suffix? 
do this new additional 9 characters fit in the max length restrictions of the resources?

Comment (felipecrescencio-cit) [reply 1-org/modules/centralized-logging/main.tf:54]: The module needs this random to generate a logging target name when the user does not provide `var.logging_target_name`.
The name that is generated will be: destination-type-prefix + random-string. i.e: ds_vib7t43r

File: 1-org/modules/centralized-logging/main.tf [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+  resource_name              = var.logging_target_type == "bigquery" ? module.destination_bigquery[0].resource_name : var.logging_target_type == "pubsub" ? module.destination_pubsub[0].resource_name : var.logging_target_type == "storage" ? module.destination_storage[0].resource_name : module.destination_logbucket[0].resource_name
+  destination_uri            = length(var.logging_destination_uri) > 0 ? var.logging_destination_uri : var.logging_target_type == "bigquery" ? module.destination_bigquery[0].destination_uri : var.logging_target_type == "pubsub" ? module.destination_pubsub[0].destination_uri : var.logging_target_type == "storage" ? module.destination_storage[0].destination_uri : module.destination_logbucket[0].destination_uri
+  create_destination         = !(length(var.logging_destination_uri) > 0)
+  logging_sink_name          = length(var.logging_sink_name) > 0 ? var.logging_sink_name : "sk-to-${local.logging_target_name_prefix}-${var.logging_destination_project_id}"
+  logging_target_name_prefix = var.logging_target_type == "bigquery" ? "ds" : var.logging_target_type == "pubsub" ? "topic" : var.logging_target_type == "storage" ? "bkt" : "logbkt"
```
Comment (daniel-cit) [inline 1-org/modules/centralized-logging/main.tf]: this could be a lookup in a map https://www.terraform.io/language/functions/lookup
it is really hard to follow all these nested inline ifs

File: 1-org/modules/centralized-logging/main.tf (hunk 5 above)
Comment (daniel-cit) [reply 1-org/modules/centralized-logging/main.tf:54]: you mean `ds-qwertyui` ?

big query dataset names do not need a random suffix, they are not global resources, but it is not a problem to have one.

you could use a different default for each type

- "ds_audit_logs_{random_string.suffix.result}"  => can use `_` but not `-`
- "bkt-${module.org_audit_logs.project_id}-audit-logs-${random_string.suffix.result}" => needs more things to prevent collision
- "tp-audit-logs-${random_string.suffix.result}"

File: 1-org/modules/centralized-logging/main.tf (hunk 6 above)
Comment (felipecrescencio-cit) [reply 1-org/modules/centralized-logging/main.tf]: Done with lookup

File: 1-org/modules/centralized-logging/main.tf [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  key_first_resource         = keys(var.resources)[0]
+  logbucket_sink_member      = { for k, v in var.resources : k => v if k != var.logging_project_key }
+  resource_name              = var.logging_target_type == "bigquery" ? module.destination_bigquery[0].resource_name : var.logging_target_type == "pubsub" ? module.destination_pubsub[0].resource_name : var.logging_target_type == "storage" ? module.destination_storage[0].resource_name : module.destination_logbucket[0].resource_name
```
Comment (daniel-cit) [inline 1-org/modules/centralized-logging/main.tf]:
```suggestion
resource_map = {
  "bigquery" = module.destination_bigquery[0].resource_name ,
  "pubsub" = module.destination_pubsub[0].resource_name,
  "storage" = module.destination_storage[0].resource_name,
}
  resource_name = lookup(resource_map, var.logging_target_type, module.destination_logbucket[0].resource_name ) 
```

does this one works?

File: 1-org/modules/centralized-logging/main.tf (hunk 4 above)
Comment (felipecrescencio-cit) [reply 1-org/modules/centralized-logging/main.tf]: Done.

File: 1-org/modules/centralized-logging/main.tf (hunk 5 above)
Comment (felipecrescencio-cit) [reply 1-org/modules/centralized-logging/main.tf:54]: Done

File: 1-org/modules/centralized-logging/main.tf (hunk 7 above)
Comment (felipecrescencio-cit) [reply 1-org/modules/centralized-logging/main.tf]: Done

**Comment by felipecrescencio-cit** [comment]
Internal review finished.
PR 781 opened in the official repo.
